data-library/library/script/nycoc_checkbook.py

```diff
@@ -0,0 +1,381 @@
+import xml.etree.ElementTree as ET
+import requests
+import pandas as pd
+import datetime as dt
+import calendar
+from pathlib import Path
+import shutil
+import time
+import random
+
+from . import df_to_tempfile
+
+from typing import Literal, TypedDict, Any
+
+api_endpoint = "https://www.checkbooknyc.com/api"
+
+TYPE_OF_DATA = Literal[
+    "Contracts",
+    "Budget",
+    "Revenue",
+    "Payroll",
+    "Spending",
+    "Spending_OGE",
+    "Contracts_OGE",
+    "Spending_NYCHA",
+    "Contracts_NYCHA",
+    "Payroll_NYCHA",
+]
+
+DEFAULT_RECORDS_FROM = "1"
+DEFAULT_MAX_RECORDS = "20000"
+
+DEFAULT_DATE_FORMAT = "%Y-%m-%d"
+DEFAULT_START_PERIOD = dt.date(2010, 1, 1).strftime(DEFAULT_DATE_FORMAT)
+DEFAULT_END_PERIOD = dt.date.today().strftime(DEFAULT_DATE_FORMAT)
+
+DEFAULT_DATA_PATH = Path(__file__).resolve().parent / "capital_spending_data"
+
+USER_AGENTS = [
+    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/109.0.0.0 Safari/537.36"
+    "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/109.0.0.0 Safari/537.36"
+    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/108.0.0.0 Safari/537.36"
+    "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/108.0.0.0 Safari/537.36"
+    "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/108.0.0.0 Safari/537.36"
+    "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/16.1 Safari/605.1.15"
+    "Mozilla/5.0 (Macintosh; Intel Mac OS X 13_1) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/16.1 Safari/605.1.15"
+]
+user_agent = random.choice(USER_AGENTS)
+
+
+class CriteriaValue(TypedDict):
+    name: str
+    type: Literal["value"]
+    value: str
+
+
+class CriteriaRange(TypedDict):
+    name: str
+    type: Literal["range"]
+    start: str
+    end: str
+
+
+class Scriptor:
+    def __init__(self, **kwargs):
+        self.__dict__.update(kwargs)
+
+    def ingest(
+        self,
+        user_agent: str | None = None,
+        type_of_data: TYPE_OF_DATA = "Spending",
+        records_from: str = DEFAULT_RECORDS_FROM,
+        max_records: str = DEFAULT_MAX_RECORDS,
+        response_columns: list[str] | None = None,
+        search_criteria: list[CriteriaValue | CriteriaRange] = [],
```

Comment:
Bit of a small note that wouldn't cause an issue here but still is good practice - lists and dicts should not have default values in python. If they're mutated (say, call `search_criteria.append(...)` at some point), that will persist while the current session is running. So the same call to a function could have different results each time

Reply:
Ah actually you do mutate it. So if there are multiple calls to ingest in one session, any changes made to search criteria will persist from first call to second.

The way to avoid this is default to None, and then declare as empty list if None within the function body